File: backend/src/main/java/com/ecomptaia/service/InventoryService.java
package com.ecomptaia.service;

import com.ecomptaia.entity.Inventory;
import com.ecomptaia.inventory.entity.InventoryItem;
import com.ecomptaia.entity.InventoryMovement;
import com.ecomptaia.repository.InventoryRepository;
import com.ecomptaia.repository.InventoryItemRepository;
import com.ecomptaia.repository.InventoryMovementRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

/**
 * Service pour la gestion des inventaires
 */
@Service
@Transactional
public class InventoryService {

    @Autowired
    private InventoryRepository inventoryRepository;

    @Autowired
    private InventoryItemRepository inventoryItemRepository;

    @Autowired
    private InventoryMovementRepository inventoryMovementRepository;

    /**
     * Créer un nouvel inventaire
     */
    public Inventory createInventory(String name, String description, Long companyId) {
        Inventory inventory = new Inventory();
        inventory.setName(name);
        inventory.setDescription(description);
        inventory.setCompanyId(companyId);
        inventory.setCreatedAt(LocalDateTime.now());
        inventory.setUpdatedAt(LocalDateTime.now());
        inventory.setIsActive(true);
        
        return inventoryRepository.save(inventory);
    }

    /**
     * Ajouter un article à l'inventaire
     */
    public InventoryItem addInventoryItem(Long inventoryId, String itemName, String description, 
                                        Integer quantity, Double unitPrice, String unit) {
        InventoryItem item = new InventoryItem();
        item.setInventoryId(inventoryId);
        item.setItemName(itemName);
        item.setDescription(description);
        item.setQuantity(quantity);
        item.setUnitPrice(BigDecimal.valueOf(unitPrice));
        item.setUnit(unit);
        item.setCreatedAt(LocalDateTime.now());
        item.setUpdatedAt(LocalDateTime.now());
        // setIsActive method doesn't exist, using isActive field directly
        // item.setActive(true); // Method does not exist
        
        return inventoryItemRepository.save(item);
    }

    /**
     * Enregistrer un mouvement d'inventaire
     */
    public InventoryMovement recordMovement(Long inventoryId, Long itemId, String movementType, 
                                          Integer quantity, String reason, Long userId) {
        InventoryMovement movement = new InventoryMovement();
        movement.setInventoryId(inventoryId);
        movement.setItemId(itemId);
        movement.setMovementType(InventoryMovement.MovementType.valueOf(movementType));
        movement.setQuantity(BigDecimal.valueOf(quantity));
        movement.setReason(reason);
        movement.setUserId(userId);
        movement.setCreatedAt(LocalDateTime.now());
        
        return inventoryMovementRepository.save(movement);
    }

    /**
     * Obtenir tous les inventaires d'une entreprise
     */
    public List<Inventory> getInventoriesByCompany(Long companyId) {
        return inventoryRepository.findByCompanyIdAndIsActiveTrue(companyId);
    }

    /**
     * Obtenir les articles d'un inventaire
     */
    public List<InventoryItem> getInventoryItems(Long inventoryId) {
        return inventoryItemRepository.findByInventoryIdAndIsActiveTrue(inventoryId);
    }

    /**
     * Obtenir les mouvements d'un inventaire
     */
    public List<InventoryMovement> getInventoryMovements(Long inventoryId) {
        return inventoryMovementRepository.findByInventoryIdOrderByCreatedAtDesc(inventoryId);
    }

    /**
     * Calculer le stock actuel d'un article
     */
    public Integer calculateCurrentStock(Long itemId) {
        List<InventoryMovement> movements = inventoryMovementRepository.findByItemIdOrderByCreatedAtAsc(itemId);
        BigDecimal stock = BigDecimal.ZERO;
        
        for (InventoryMovement movement : movements) {
            if ("IN".equals(movement.getMovementType().toString())) {
                stock = stock.add(movement.getQuantity());
            } else if ("OUT".equals(movement.getMovementType().toString())) {
                stock = stock.subtract(movement.getQuantity());
            }
        }
        
        return stock.intValue();
    }

    /**
     * Mettre à jour la quantité d'un article
     */
    public InventoryItem updateItemQuantity(Long itemId, Integer newQuantity) {
        Optional<InventoryItem> itemOpt = inventoryItemRepository.findById(itemId);
        if (itemOpt.isPresent()) {
            InventoryItem item = itemOpt.get();
            item.setQuantity(newQuantity);
            item.setUpdatedAt(LocalDateTime.now());
            return inventoryItemRepository.save(item);
        }
        return null;
    }

    /**
     * Désactiver un inventaire
     */
    public void deactivateInventory(Long inventoryId) {
        Optional<Inventory> inventoryOpt = inventoryRepository.findById(inventoryId);
        if (inventoryOpt.isPresent()) {
            Inventory inventory = inventoryOpt.get();
            inventory.setIsActive(false);
            inventory.setUpdatedAt(LocalDateTime.now());
            inventoryRepository.save(inventory);
        }
    }

    /**
     * Obtenir les statistiques d'inventaire
     */
    public Object getInventoryStatistics(Long companyId) {
        List<Inventory> inventories = getInventoriesByCompany(companyId);
        
        int totalCount = inventories.size();
        int activeCount = (int) inventories.stream().filter(Inventory::getIsActive).count();
        
        return new Object() {
            public final int totalInventories = totalCount;
            public final int activeInventories = activeCount;
            public final LocalDateTime lastUpdate = LocalDateTime.now();
            
            @Override
            public String toString() {
                return String.format("InventoryStats{total=%d, active=%d, updated=%s}", 
                    totalInventories, activeInventories, lastUpdate.toString());
            }
        };
    }
}

File: frontend/src/app/services/inventory.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Inventory {
  id: number;
  name: string;
  description: string;
  companyId: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface InventoryItem {
  id: number;
  inventoryId: number;
  itemName: string;
  description: string;
  quantity: number;
  unitPrice: number;
  unit: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface InventoryMovement {
  id: number;
  inventoryId: number;
  itemId: number;
  movementType: string;
  quantity: number;
  reason: string;
  userId: number;
  createdAt: string;
}

@Injectable({
  providedIn: 'root'
})
export class InventoryService {
  private apiUrl = `${environment.apiUrl}/api/inventory`;

  constructor(private http: HttpClient) { }

  /**
   * Créer un nouvel inventaire
   */
  createInventory(name: string, description: string, companyId: number): Observable<Inventory> {
    return this.http.post<Inventory>(`${this.apiUrl}/inventories`, {
      name,
      description,
      companyId
    });
  }

  /**
   * Ajouter un article à l'inventaire
   */
  addInventoryItem(inventoryId: number, itemName: string, description: string, 
                  quantity: number, unitPrice: number, unit: string): Observable<InventoryItem> {
    return this.http.post<InventoryItem>(`${this.apiUrl}/items`, {
      inventoryId,
      itemName,
      description,
      quantity,
      unitPrice,
      unit
    });
  }

  /**
   * Enregistrer un mouvement d'inventaire
   */
  recordMovement(inventoryId: number, itemId: number, movementType: string, 
                quantity: number, reason: string, userId: number): Observable<InventoryMovement> {
    return this.http.post<InventoryMovement>(`${this.apiUrl}/movements`, {
      inventoryId,
      itemId,
      movementType,
      quantity,
      reason,
      userId
    });
  }

  /**
   * Obtenir tous les inventaires d'une entreprise
   */
  getInventoriesByCompany(companyId: number): Observable<Inventory[]> {
    return this.http.get<Inventory[]>(`${this.apiUrl}/inventories/company/${companyId}`);
  }

  /**
   * Obtenir les articles d'un inventaire
   */
  getInventoryItems(inventoryId: number): Observable<InventoryItem[]> {
    return this.http.get<InventoryItem[]>(`${this.apiUrl}/items/inventory/${inventoryId}`);
  }

  /**
   * Obtenir les mouvements d'un inventaire
   */
  getInventoryMovements(inventoryId: number): Observable<InventoryMovement[]> {
    return this.http.get<InventoryMovement[]>(`${this.apiUrl}/movements/inventory/${inventoryId}`);
  }

  /**
   * Calculer le stock actuel d'un article
   */
  calculateCurrentStock(itemId: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/stock/item/${itemId}`);
  }

  /**
   * Mettre à jour la quantité d'un article
   */
  updateItemQuantity(itemId: number, newQuantity: number): Observable<InventoryItem> {
    return this.http.put<InventoryItem>(`${this.apiUrl}/items/${itemId}/quantity`, {
      quantity: newQuantity
    });
  }

  /**
   * Désactiver un inventaire
   */
  deactivateInventory(inventoryId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/inventories/${inventoryId}`);
  }

  /**
   * Obtenir les statistiques d'inventaire
   */
  getInventoryStatistics(companyId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/statistics/company/${companyId}`);
  }

  /**
   * Rechercher des inventaires par nom
   */
  searchInventoriesByName(name: string, companyId: number): Observable<Inventory[]> {
    return this.http.get<Inventory[]>(`${this.apiUrl}/inventories/search?name=${name}&companyId=${companyId}`);
  }

  /**
   * Rechercher des articles par nom
   */
  searchItemsByName(itemName: string, inventoryId: number): Observable<InventoryItem[]> {
    return this.http.get<InventoryItem[]>(`${this.apiUrl}/items/search?itemName=${itemName}&inventoryId=${inventoryId}`);
  }
}

